### ISSUE TITLE
Refactoring of `smithy-rules-engine` for code generation

### ISSUE THREAD

**Issue by kstich** [awslabs/smithy-rs#2844]
The package `smithy-rules-engine` has undergone a major refactoring in an effort to improve validation, separate AWS and non-AWS concerns, add a specification, and better align with Smithy's coding standards. [A code-stable pull request is available here](https://github.com/smithy-lang/smithy/pull/1855). General notes are provided in individual commit messages. The format of the rules documents have _**not**_ changed, meaning a successful migration to the refactored codebase will involve _**no changes**_ to code generated for an SDK client.

The Smithy team plans to release this update on August 15th, 2023. Please reach out to us for any additional information or concerns.

**Comment by kstich** [comment]
To account for upcoming traits to simplify the AWS endpoints modeling story for service teams, the AWS-specific components in this work have moved from `smithy-aws-rules-engine` to `smithy-aws-endpoints`. No other changes to the code within the package has been made.

To accomodate this change, we're pushing the planned release of the refactoring to August 22nd, 2023.

**Comment by rcoh** [comment]
That sounds good to me—I've verified that it is working in a branch. Note that I've pulled the tests out of Smithy and [vended them in our own repo](https://github.com/awslabs/smithy-rs/tree/8ac05483c5a759eb3e926a60ed56d1cec883c9a6/codegen-client/src/test/resources/META-INF/smithy/endpoint-tests). This isn't ideal because it means we won't keep up with any updates to Smithy-provided rules language tests. It would be ideal if Smithy could vend these .smithy files to use for SDKs to run their own tests.
